govcd/metadata_v2.go

```diff
@@ -0,0 +1,754 @@
+func (disk *Disk) DeleteMetadataEntryWithDomain(key string, isSystem bool) error {
+	return deleteMetadataAndWait(disk.client, disk.Disk.HREF, key, isSystem)
+}
+
+// DeleteMetadataEntryWithDomain deletes OrgVDCNetwork metadata associated to the input key and waits for the task to finish.
+// Note: Requires system administrator privileges.
+func (orgVdcNetwork *OrgVDCNetwork) DeleteMetadataEntryWithDomain(key string, isSystem bool) error {
+	return deleteMetadataAndWait(orgVdcNetwork.client, getAdminURL(orgVdcNetwork.OrgVDCNetwork.HREF), key, isSystem)
+}
+
+// DeleteMetadataEntryWithDomain deletes CatalogItem metadata associated to the input key and waits for the task to finish.
+func (catalogItem *CatalogItem) DeleteMetadataEntryWithDomain(key string, isSystem bool) error {
+	return deleteMetadataAndWait(catalogItem.client, catalogItem.CatalogItem.HREF, key, isSystem)
+}
+
+// DeleteMetadataEntryWithDomain deletes OpenApiOrgVdcNetwork metadata associated to the input key and waits for the task to finish.
+// Note: It doesn't delete metadata from networks that belong to a VDC Group.
+// TODO: This function is currently using XML API underneath as OpenAPI metadata is still not supported.
+func (openApiOrgVdcNetwork *OpenApiOrgVdcNetwork) DeleteMetadataEntryWithDomain(key string, isSystem bool) error {
+	href := fmt.Sprintf("%s/admin/network/%s", openApiOrgVdcNetwork.client.VCDHREF.String(), extractUuid(openApiOrgVdcNetwork.OpenApiOrgVdcNetwork.ID))
+	task, err := deleteMetadata(openApiOrgVdcNetwork.client, href, key, isSystem)
+	if err != nil {
+		return err
+	}
+	return task.WaitTaskCompletion()
+}
+
+// ------------------------------------------------------------------------------------------------
+// Generic private functions
+// ------------------------------------------------------------------------------------------------
+
+// getMetadata is a generic function to retrieve metadata from VCD
+func getMetadata(client *Client, requestUri string) (*types.Metadata, error) {
+	metadata := &types.Metadata{}
+
+	_, err := client.ExecuteRequest(requestUri+"/metadata/", http.MethodGet, types.MimeMetaData, "error retrieving metadata: %s", nil, metadata)
+	return metadata, err
+}
+
+// addMetadata adds metadata to an entity.
+// If the metadata entry is of the SYSTEM domain (isSystem=true), one can set different types of Visibility:
+// types.MetadataReadOnlyVisibility, types.MetadataHiddenVisibility but NOT types.MetadataReadWriteVisibility.
+// If the metadata entry is of the GENERAL domain (isSystem=false), visibility is always types.MetadataReadWriteVisibility.
+// In terms of typedValues, that must be one of:
+// types.MetadataStringValue, types.MetadataNumberValue, types.MetadataDateTimeValue and types.MetadataBooleanValue.
+func addMetadata(client *Client, requestUri, key, value, typedValue, visibility string, isSystem bool) (Task, error) {
+
+	if isSystem && visibility == types.MetadataReadWriteVisibility {
```

Comment:
While self-reviewing, I was thinking that having this conditional may be adding too much logic/awareness to this function. If someone introduces a wrong combination of `visibility` and `isSystem`, probably we should let VCD return an error, just in case something changes in future in VCD side.

Let me know your thoughts.

Reply:
I am all for VCD validation if the error coming from VCD is human readable and hints what is wrong. This is usually the approach we take to avoid over-validation.

Reply:
Yes, strategically, we always let CD API validate and add our own validation only where it's absolutely needed.

Reply:
If the error is an internal system error (with usually too much information and the real cause buried deep down) we should avoid exposing users to it

Reply:
I re-visited this use-case and the error coming from VCD is like this:
`API Error: 500: [ 83cca9ab-826f-406b-a641-de907cf8a3fe ] visibility`
What do you think? Probably it makes sense in this case, doesn't it?

Reply:
I actually was playing a bit with it in Terraform. What I didn't like is that I have like 10 metadata items and all I get is that error. A bit ugly that I can't know which item throws this error. I can still live with it, but a bit more information could be nice. 
<img width="1254" alt="image" src="https://user-images.githubusercontent.com/15804230/197758010-42253fc8-f939-4ac0-a600-91654092d848.png">

Reply:
Ah, didn't think about it, but my error message would be useless even in that scenario. I will change it.

Reply:
I refactored the code, instead of "being too smart", I just beautified the error message for this unique case.

Reply:
> 	// Workaround for ugly error returned by VCD: "API Error: 500: [ <uuid> ] visibility"
>	if err != nil && strings.HasSuffix(err.Error(), "visibility") {
> 		err = fmt.Errorf("error adding metadata with key %s: visibility cannot be %s when domain is %s", key, visibility, domain)

May there be another CD error with "visibility" in it which we don't expect? Just thinking to suggest adding the original err message at the very end of this custom message for completeness.

Reply:
Updated 👍